core/src/main/java/org/infinispan/util/concurrent/CompletionStages.java

```diff
@@ -0,0 +1,201 @@
+   }
+
+   /**
+    * Shorthand for <code>CompletionStages.join(CompletionStages.allOf(stages))</code> but doesn't return a value
+    * @param stages
+    * @throws CompletionException if this future completed exceptionally or a completion computation threw an exception
+    */
+   public static void await(CompletionStage<?>... stages) {
+      join(allOf(stages));
+   }
+
+   /**
+    * Returns a CompletableStage that completes when both of the provides CompletionStages complete. This method
+    * may choose to return either of the argument if the other is complete or a new instance completely.
+    * @param first the first CompletionStage
+    * @param second the second CompletionStage
+    * @return a CompletionStage that is complete when both of the given CompletionStages complete
+    */
+   public static CompletionStage<Void> allOf(CompletionStage<Void> first, CompletionStage<Void> second) {
+      if (!isCompleteSuccessfully(first)) {
+         if (isCompleteSuccessfully(second)) {
+            return first;
+         } else {
+            return CompletionStages.aggregateCompletionStage().dependsOn(first).dependsOn(second).freeze();
+         }
+      }
+      return second;
+   }
+
+   /**
+    * Returns a CompletionStage that completes when all of the provided stages complete, either normally or via
+    * exception. If one or more states complete exceptionally the returned CompletionStage will complete with the
+    * exception of one of these. If no CompletionStages are provided, returns a CompletionStage completed with the value
+    * null.
+    * @param stages the CompletionStages
+    * @return a CompletionStage that is completed when all of the given CompletionStages complete
+    */
+   public static CompletionStage<Void> allOf(CompletionStage<?>... stages) {
+      AggregateCompletionStage<Void> aggregateCompletionStage = null;
+      for (CompletionStage<?> stage : stages) {
+         if (!isCompleteSuccessfully(stage)) {
+            if (aggregateCompletionStage == null) {
+               aggregateCompletionStage = CompletionStages.aggregateCompletionStage();
+            }
+            aggregateCompletionStage.dependsOn(stage);
+         }
+      }
+
+      return aggregateCompletionStage != null ? aggregateCompletionStage.freeze() : CompletableFutures.completedNull();
+   }
+
+   private static class VoidAggregateCompletionStage extends AbstractAggregateCompletionStage<Void> {
+      @Override
+      Void getValue() {
+         return null;
+      }
+   }
+
+   private static class ValueAggregateCompletionStage<R> extends AbstractAggregateCompletionStage<R> {
+      private final R value;
+
+      private ValueAggregateCompletionStage(R value) {
+         this.value = value;
+      }
+
+      @Override
+      R getValue() {
+         return value;
+      }
+   }
+
+   // This class implements BiConsumer to avoid additional object/lambda allocation per instance
```

Comment:
Why not javadoc?

Reply:
Wasn't public, so I wasn't too worried about it. But I can add it in if you want.